File: moqt/internal/message/subscribe_end.go
package message

import (
	"io"
)

// SubscribeEndMessage is sent by the publisher to signal that no group
// after the given sequence will be produced.
//
// Wire format:
//
//	SUBSCRIBE_END Message {
//	  Type (i) = 0x1
//	  Message Length (i)
//	  Group (i)
//	}
type SubscribeEndMessage struct {
	// Group is the absolute sequence number of the last group that may be
	// delivered, inclusive (plain absolute sequence, not the +1 form).
	Group uint64
}

func (sem SubscribeEndMessage) Len() int {
	return VarintLen(sem.Group)
}

func (sem SubscribeEndMessage) Encode(w io.Writer) error {
	msgLen := sem.Len()
	b := make([]byte, 0, msgLen+VarintLen(uint64(msgLen)))

	b, _ = WriteMessageLength(b, uint64(msgLen))
	b, _ = WriteVarint(b, sem.Group)

	_, err := w.Write(b)

	return err
}

func (sem *SubscribeEndMessage) Decode(src io.Reader) error {
	size, err := ReadMessageLength(src)
	if err != nil {
		return err
	}

	if size > MaxMessageSize {
		return ErrMessageTooLarge
	}

	b := make([]byte, size)
	_, err = io.ReadFull(src, b)
	if err != nil {
		return err
	}

	num, n, err := ReadVarint(b)
	if err != nil {
		return err
	}
	sem.Group = num
	b = b[n:]

	if len(b) != 0 {
		return ErrMessageTooShort
	}

	return nil
}

File: moq-web/src/internal/message/subscribe_end.ts
import type { Reader, Writer } from "@okdaichi/golikejs/io";
import { MessageDecoder, MessageEncoder, readFull, readVarint } from "./message.ts";

export interface SubscribeEndMessageInit {
	group?: number;
}

/**
 * SUBSCRIBE_END message (type 0x1), sent by the publisher to signal that
 * no group after the given sequence will be produced. The type varint is
 * written by the caller before `encode` / consumed before `decode`.
 *
 * ```text
 * SUBSCRIBE_END Message {
 *   Type (i) = 0x1
 *   Message Length (i)
 *   Group (i)
 * }
 * ```
 */
export class SubscribeEndMessage {
	/** Absolute sequence of the last group that may be delivered (inclusive). */
	group: number;

	constructor(init: SubscribeEndMessageInit = {}) {
		this.group = init.group ?? 0;
	}

	/**
	 * Encodes the message to the writer.
	 */
	async encode(w: Writer): Promise<Error | undefined> {
		const e = new MessageEncoder();
		e.varint(this.group);
		const [, err] = await w.write(e.frame());
		return err;
	}

	/**
	 * Decodes the message from the reader.
	 */
	async decode(r: Reader): Promise<Error | undefined> {
		const [msgLen, , err1] = await readVarint(r);
		if (err1) return err1;

		const buf = new Uint8Array(msgLen);
		const [, err2] = await readFull(r, buf);
		if (err2) return err2;

		const d = new MessageDecoder(buf);
		this.group = d.varint();

		return undefined;
	}
}
